drivers/crypto/bflb-seceng.c

```diff
@@ -0,0 +1,840 @@
+	struct bflb_seceng *seceng;
+};
+
+struct regmap_config bflb_seceng_regmap_config = {
+	.reg_bits = 32,
+	.val_bits = 32,
+	.reg_stride = 4,
+	.cache_type = REGCACHE_FLAT,
+	.max_register = 512 * sizeof(u32),
+	.num_reg_defaults_raw = 512,
+	.use_relaxed_mmio = true,
+	.use_raw_spinlock = true,
+};
+
+static struct bflb_seceng *bflb_seceng_dev;
+
+static void bflb_seceng_trng_wait_ready(void)
+{
+	while (readl_relaxed(bflb_seceng_dev->base + REG_SECENG_TRNG_0_CTRL_0) &
+			REG_SECENG_TRNG_0_CTRL_0_BUSY) {
+		dev_dbg(bflb_seceng_dev->dev, "Waiting for TRNG ready");
+		schedule();
+	}
+}
+
+static void bflb_seceng_trng_init(void)
+{
+	regmap_update_bits
+	(
+		bflb_seceng_dev->map,
+		REG_SECENG_TRNG_0_CTRL_0,
+		REG_SECENG_TRNG_0_CTRL_0_EN | REG_SECENG_TRNG_0_CTRL_0_INT_CLR_1T,
+		FIELD_PREP(REG_SECENG_TRNG_0_CTRL_0_EN, 1) |
+				FIELD_PREP(REG_SECENG_TRNG_0_CTRL_0_INT_CLR_1T, 1)
+	);
+
+	bflb_seceng_trng_wait_ready();
+}
+
+static void bflb_seceng_trng_refresh(void)
+{
+	regmap_update_bits
+	(
+		bflb_seceng_dev->map,
+		REG_SECENG_TRNG_0_CTRL_0,
+		REG_SECENG_TRNG_0_CTRL_0_DOUT_CLR_1T,
+		FIELD_PREP(REG_SECENG_TRNG_0_CTRL_0_DOUT_CLR_1T, 1)
+	); //Clear DOUT
+
+	bflb_seceng_trng_wait_ready();
+
+	regmap_update_bits
+	(
+		bflb_seceng_dev->map,
+		REG_SECENG_TRNG_0_CTRL_0,
+		REG_SECENG_TRNG_0_CTRL_0_DOUT_CLR_1T,
+		FIELD_PREP(REG_SECENG_TRNG_0_CTRL_0_DOUT_CLR_1T, 0)
+	); //Reset clear DOUT
+
+	regmap_update_bits
+	(
+		bflb_seceng_dev->map,
+		REG_SECENG_TRNG_0_CTRL_0,
+		REG_SECENG_TRNG_0_CTRL_0_TRIG_1T,
+		FIELD_PREP(REG_SECENG_TRNG_0_CTRL_0_TRIG_1T, 1)
+	); //Force TRNG refresh
+
+	bflb_seceng_trng_wait_ready();
+
+	regmap_update_bits
+	(
+		bflb_seceng_dev->map,
+		REG_SECENG_TRNG_0_CTRL_0,
+		REG_SECENG_TRNG_0_CTRL_0_INT_CLR_1T,
+		FIELD_PREP(REG_SECENG_TRNG_0_CTRL_0_INT_CLR_1T, 1)
+	); //Clear INT
+
+	dev_dbg(bflb_seceng_dev->dev, "Refreshed TRNG");
+}
+
+static unsigned int bflb_seceng_read_rng_dout(struct bflb_seceng *seceng, unsigned int doutreg)
+{
+	unsigned int selectedreg = 0;
+
+	switch (doutreg) {
+	case 0:
+		selectedreg = REG_SECENG_TRNG_0_DOUT_0;
+		break;
+
+	case 1:
+		selectedreg = REG_SECENG_TRNG_0_DOUT_1;
+		break;
+
+	case 2:
+		selectedreg = REG_SECENG_TRNG_0_DOUT_2;
+		break;
+
+	case 3:
+		selectedreg = REG_SECENG_TRNG_0_DOUT_3;
+		break;
+
+	case 4:
+		selectedreg = REG_SECENG_TRNG_0_DOUT_4;
+		break;
+
+	case 5:
+		selectedreg = REG_SECENG_TRNG_0_DOUT_5;
+		break;
+
+	case 6:
+		selectedreg = REG_SECENG_TRNG_0_DOUT_6;
+		break;
+
+	case 7:
+		selectedreg = REG_SECENG_TRNG_0_DOUT_7;
+		break;
+
+	default: return -EINVAL;
+	}
+
+	return readl_relaxed(seceng->base + selectedreg);
+}
+
+static int bflb_seceng_trng_crypto_read(struct bflb_seceng *seceng, u8 *dstn,
+		unsigned int dlen)
+{
+	unsigned int written = 0;
+	unsigned int doutreg = 8;
+	u32 val;
+
+	mutex_lock(&bflb_seceng_dev->lock);
+
+	dev_dbg(seceng->dev, "Starting TRNG crypto read for %u bytes...", dlen);
+
+	do {
+		if (doutreg >= 8) {
```

Comment:
This complicated loop/math can be simplified if you refactor into the following functions:
*crypto_read_multiple() - this function: calls crypto_read8() in a loop to fill in the buffer one byte at a time
*crypto_read8() - calls *crypto_read32() to get 32-bits of randomness and returns 8 bits of it.  remembers previous cached 32-bit value and returns next 8 bits of it until empty and then refreshes with another call to *crypto_read32()
*crypto_read32() - returns the next 32-bit dout value, and calls refresh() when there is no more randomness